scripts/system/create/edit.js

```diff
@@ -2489,6 +2489,35 @@ var PropertiesTool = function (opts) {
                     pushCommandForSelections();
                     selectionManager._update(false, this);
                 }
+            } else if (data.action === "resetToNaturalPivot") {
+                if (selectionManager.hasSelection()) {
+                    selectionManager.saveProperties();
+                    for (i = 0; i < selectionManager.selections.length; i++) {
+                        properties = selectionManager.savedProperties[selectionManager.selections[i]];
+                        var naturalDimensions = properties.naturalDimensions;
+
+                        // If any of the natural dimensions are not 0, set the entity's pivot according to its natural dimensions and natural position.
+                        // In this way, a model entity will match the pivot of its model.
+                        if (properties.type === "Model" && naturalDimensions.x === 0 && naturalDimensions.y === 0 &&
+                            naturalDimensions.z === 0) {
+                            Window.notifyEditError("Cannot apply the model's pivot to the entity: Model URL" +
+                                " is invalid or the model has not yet been loaded.");
+                        } else {
+                            var naturalMinimumExtent = Vec3.subtract(properties.naturalPosition, Vec3.multiply(naturalDimensions, 0.5));
+                            var negativeNaturalMinimumExtent = Vec3.multiply(naturalMinimumExtent, -1);
+
+                            var invNaturalDimensions = { x: 1.0 / naturalDimensions.x, y: 1.0 / naturalDimensions.y, z: 1.0 / naturalDimensions.z };
+
+                            var naturalPivot = Vec3.multiplyVbyV(negativeNaturalMinimumExtent, invNaturalDimensions);
+
+                            Entities.editEntity(selectionManager.selections[i], {
+                                registrationPoint: naturalPivot
```

Comment:
Instead of setting the registration point and removing the range etc, could we just set the position here?

Reply:
Hi, do you mean set the entity's position?  That's not the desired outcome; it's definitely the pivot we want to control.

Normally, when we set an entity's pivot, it's stored in the registration point.  Could you explain why it should be any different when we want the entity's pivot to match its model's pivot?
We could bypass the registration point and do something else for that case, but what would be the advantage?  It would seem to me like adding another pivot system alongside the one we have.

Thanks for the feedback :)

Reply:
I’m concerned that this is overloading the use of registration point and that doing it this way will have unintended consequences for culling, physics, picking, etc, especially now that registration point isn’t clamped.  I also think this should be a top level entity property, so that people can properly toggle it.  This is an awesome feature and one that I expect lots of people will use, so it seems important to implement it in a future-proof way.  For example, this implementation only works using the built in Create app, but other scripters still have no way of doing it automatically.

Reply:
I see, thanks, those are good points.  So far I've been considering the registration point as just the pivot (local origin) by another name, but what's it actually for?  It's presented as a pivot in the Create app, and works like one, except that it's clamped within the geometry bounds and I didn't know why.  I thought it was most likely just clamped because no-one had tested it outside that range yet.  The tests I've done for culling, collision and picking have shown no problems with faraway pivots, but there could well be things I don't know about.

Reply:
It has a similar effect, but I think it’s not a good idea to overload it like this.  The registration point is relative to the dimensions of the entity.  If someone models a scene and exports each model as a separate entity (with their own pivots), setting the pivots like this will of course work and position them properly in Interface...but the actual values of the registrations points will be meaningless.  Ideally the pivots would be presented in world coordinates, or not at all, and you’d just toggle the “Respect Original Pivot” button and not worry about the values at all.